### ISSUE TITLE
3.8.4: Unable to load the EventMachine C extension

### ISSUE THREAD

**Issue by Birch-san** [envygeeks/jekyll-docker#241]
- [x] I tried updating to the latest version
  - [x] I can't, there is an issue
- [x] I am on macOS 10.14
- [x] I'm on Docker
  - [ ] I understand Docker may be unsupported

## Description

The run script I've used successfully on 3.8.3, fails under 3.8.4.

Encounters:

- `Unable to load the EventMachine C extension`
- `Error loading shared library libssl.so.44: No such file or directory`.

So, we need to install libssl. Or switch to the pure-Ruby eventmachine.

## Steps

- Clone [birchlabs-site, branch `jekyll-3.8.4`](https://github.com/Birch-san/birchlabs-site/blob/jekyll-3.8.4/run.sh).
- Invoke [run.sh](https://github.com/Birch-san/birchlabs-site/blob/master/run.sh)

```bash
git clone -b jekyll-3.8.4 https://github.com/Birch-san/birchlabs-site.git
cd birchlabs-site
./run.sh
```

## Output

- [Default verbosity](https://gist.github.com/Birch-san/54e3c0a8d0e2ab7e8f75f9842bef7bef)
- [`-e DEBUG=true` + `--verbose`](https://gist.github.com/Birch-san/c5144468a5a4d4843c9c271d07e8f90d)

```
Unable to load the EventMachine C extension; To use the pure-ruby reactor, require 'em/pure_ruby'
Traceback (most recent call last):
	20: from /usr/local/bundle/bin/jekyll:29:in `<main>'
	19: from /usr/local/bundle/bin/jekyll:29:in `load'
	18: from /usr/local/bundle/gems/jekyll-3.7.0/exe/jekyll:15:in `<top (required)>'
	17: from /usr/local/bundle/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
…
	 5: from /usr/local/bundle/gems/jekyll-3.7.0/lib/jekyll/commands/serve/live_reload_reactor.rb:4:in `<top (required)>'
	 4: from /usr/local/bundle/gems/jekyll-3.7.0/lib/jekyll/commands/serve/live_reload_reactor.rb:4:in `require'
	 3: from /usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket.rb:3:in `<top (required)>'
	 2: from /usr/local/bundle/gems/em-websocket-0.5.1/lib/em-websocket.rb:3:in `require'
	 1: from /usr/local/bundle/gems/eventmachine-1.2.5/lib/eventmachine.rb:8:in `<top (required)>'
/usr/local/bundle/gems/eventmachine-1.2.5/lib/eventmachine.rb:8:in `require': Error loading shared library libssl.so.44: No such file or directory (needed by /usr/local/bundle/gems/eventmachine-1.2.5/lib/rubyeventmachine.so) - /usr/local/bundle/gems/eventmachine-1.2.5/lib/rubyeventmachine.so (LoadError)
```

## Expected

Here's the Jekyll 3.8.3 behaviour, with the same run script:

```
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-musl]
Configuration file: /srv/jekyll/_config.yml
            Source: /srv/jekyll
       Destination: /dist
 Incremental build: disabled. Enable with --incremental
      Generating... 
                    done in 8.516 seconds.
 Auto-regeneration: enabled for '/srv/jekyll'
LiveReload address: http://0.0.0.0:35729
    Server address: http://0.0.0.0:4000/
  Server running... press ctrl-c to stop.
```

**Comment by envygeeks** [comment]
I’m curious how it built at all without libssl-dev unless something is providing it without an explicit dependency. Exp Upstream to us but downstream Ruby maybe. I’m not seeing anything diverging in your gemfile so that’s curious. None the less, it’ll get added so livereload is not broken!

**Comment by Birch-san** [comment]
For reference, here's what SSL packages are installed:

```bash
docker run --rm jekyll/jekyll:3.8.4 sh -c 'apk info | grep ssl' 
WARNING: Ignoring APKINDEX.adfa7ceb.tar.gz: No such file or directory
WARNING: Ignoring APKINDEX.efaa1f73.tar.gz: No such file or directory
libressl2.7-libcrypto
libressl2.7-libssl
libressl2.7-libtls
ssl_client
libressl
libssl1.0
```

**Comment by Birch-san** [comment]
I don't think there's any apk package that will give us libssl.so.44:

https://pkgs.alpinelinux.org/contents?file=libssl.so.44&path=&name=&arch=x86_64

**Comment by Birch-san** [comment]
I figured out the problem.
I'd been volume-mounting a gem cache which bundler had created under 3.8.3. This also meant that bundler installed native extensions during that time; it compiled eventmachine's native extensions, linking to whatever libssl had been available at the time.

The solution was to have bundler install eventmachine again, and rebuild native extensions against my newer libraries. In short: I deleted the gem cache and that fixed it.
